amm/contracts/router_v2/lib.rs

```diff
@@ -0,0 +1,639 @@
+                to,
+            )?;
+            Ok(())
+        }
+
+        /// Computes the amounts of tokens that have to be supplied
+        /// at each step of the exchange `path`, to get exactly `amount_out` of `token_out`
+        /// tokens at the end of the swaps.
+        fn calculate_amounts_in(
+            &self,
+            amount_out: u128,
+            valid_path: &[ValidStep],
+            token_out: AccountId,
+        ) -> Result<Vec<u128>, RouterV2Error> {
+            let mut amounts = vec![0; valid_path.len() + 1];
+            amounts[valid_path.len() - 1] = amount_out;
+            amounts[valid_path.len() - 2] = valid_path[valid_path.len() - 2].pool.get_amount_in(
+                valid_path[valid_path.len() - 2].token_in,
+                token_out,
+                amount_out,
+            )?;
+            for i in (0..valid_path.len() - 1).rev() {
+                amounts[i] = valid_path[i].pool.get_amount_in(
+                    valid_path[i].token_in,
+                    valid_path[i + 1].token_in,
+                    amounts[i + 1],
+                )?;
+            }
+
+            Ok(amounts)
+        }
+
+        /// Computes swap token amounts over the given path.
+        ///
+        /// Returns list of swap outcomes along the path.
+        fn calculate_amounts_out(
+            &self,
+            amount_in: u128,
+            valid_path: &[ValidStep],
+            token_out: AccountId,
+        ) -> Result<Vec<u128>, RouterV2Error> {
+            let mut amounts = Vec::with_capacity(valid_path.len() + 1);
+            amounts.push(amount_in);
+            for i in 0..valid_path.len() - 1 {
+                amounts.push(valid_path[i].pool.get_amount_out(
+                    valid_path[i].token_in,
+                    valid_path[i + 1].token_in,
+                    amounts[i],
+                )?);
+            }
+            amounts.push(valid_path[valid_path.len() - 1].pool.get_amount_out(
+                valid_path[valid_path.len() - 1].token_in,
+                token_out,
+                amounts[valid_path.len() - 1],
+            )?);
+
+            Ok(amounts)
+        }
+
+        /// Checks if the path is valid.
+        ///
+        /// Returns valid path
+        fn validate_path(
+            &self,
+            path: &[Step],
+            token_out: AccountId,
+        ) -> Result<Vec<ValidStep>, RouterV2Error> {
+            ensure!(!path.is_empty(), RouterV2Error::InvalidPath);
+            let mut valid_path = Vec::with_capacity(path.len());
+            for i in 0..(path.len() - 1) {
+                ensure!(
+                    path[i].token_in != path[i + 1].token_in,
+                    RouterV2Error::IdenticalAddresses
+                );
+                valid_path.push(ValidStep {
+                    token_in: path[i].token_in,
+                    pool: self
+                        .get_pool(path[i].pool_id, (path[i].token_in, path[i + 1].token_in))?,
```

Comment:
I agree. The reason for storing `Pair`s by their `(token_0, token_1)` instead of `AccountId` is to allow creating v2 pools when calling `add_pair_liqudity(token_0, token_1, ...)` - we need to check first if a v2 `Pair` exists for given `(token_0, token_1)`.

We could ofc store `Pair`s additionally by their `AccountId` in a separate Mapping, or together with `StablePools` in a `Mapping<AccountId, Pool>` where `Pool` is enum `Pool::Pair(Pair(...)) Pool::StablePool(StablePool(...))` but I think it would make lots of redundant data stored in the contract.

Reply:
Makes sense, let's leave it then as it is